test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs

```diff
@@ -0,0 +1,264 @@
+
+                solutionService.Save();
+                project.Build();
+                Assert.True(VisualStudio.HasNoErrorsInErrorList());
+                Assert.True(VisualStudio.HasNoErrorsInOutputWindows());
+
+                nugetConsole.Clear();
+            }
+        }
+
+        [CIOnlyNuGetWpfTheory]
+        [InlineData(ProjectTemplate.ClassLibrary)]
+        [InlineData(ProjectTemplate.NetCoreConsoleApp)]
+        [InlineData(ProjectTemplate.NetStandardClassLib)]
+        public void UpdateUnsignedPackageToSignedVersionFromPMC(ProjectTemplate projectTemplate)
+        {
+            var packageVersion09 = "0.9.0";
+            var signedPackage = _fixture.SignedTestPackage;
+
+            using (var pathContext = new SimpleTestPathContext())
+            {
+                // Arrange
+                EnsureVisualStudioHost();
+                var solutionService = VisualStudio.Get<SolutionService>();
+
+                solutionService.CreateEmptySolution("TestSolution", pathContext.SolutionRoot);
+                var project = solutionService.AddProject(ProjectLanguage.CSharp, projectTemplate, ProjectTargetFramework.V46, "TestProject");
+                project.Build();
+
+                Utils.CreatePackageInSource(pathContext.PackageSource, signedPackage.Id, packageVersion09);
+                SimpleTestPackageUtility.CreatePackages(pathContext.PackageSource, signedPackage);
+
+                var nugetTestService = GetNuGetTestService();
+                Assert.True(nugetTestService.EnsurePackageManagerConsoleIsOpen());
+
+                var nugetConsole = nugetTestService.GetPackageManagerConsole(project.UniqueName);
+
+                Assert.True(nugetConsole.InstallPackageFromPMC(signedPackage.Id, packageVersion09));
+                Assert.True(Utils.IsPackageInstalled(nugetConsole, project.FullPath, signedPackage.Id, packageVersion09));
+                project.Build();
+
+                Assert.True(nugetConsole.UpdatePackageFromPMC(signedPackage.Id, signedPackage.Version));
+                Assert.True(Utils.IsPackageInstalled(nugetConsole, project.FullPath, signedPackage.Id, signedPackage.Version));
+                Assert.False(Utils.IsPackageInstalled(nugetConsole, project.FullPath, signedPackage.Id, packageVersion09));
+                project.Build();
+
+                Assert.True(VisualStudio.HasNoErrorsInErrorList());
+                Assert.True(VisualStudio.HasNoErrorsInOutputWindows());
+
+                nugetConsole.Clear();
+                solutionService.Save();
+            }
+        }
+
+        [CIOnlyNuGetWpfTheory]
+        [InlineData(ProjectTemplate.ClassLibrary)]
+        [InlineData(ProjectTemplate.NetCoreConsoleApp)]
+        [InlineData(ProjectTemplate.NetStandardClassLib)]
+        public void DowngradeSignedPackageToUnsignedVersionFromPMC(ProjectTemplate projectTemplate)
```

Comment:
Why are we allowing to update from a signed package to unsigned package? I don't see any logical reason for this.

Reply:
It's not ideal, but that's the current state of the product and it's fine to capture current non-ideal product state in a test.  The current behavior even with unsigned packages is "first package source wins."  I believe signed packages should have special treatment in that they should not be silently updated to an unsigned version or a different signer, but also depending on the current effective client policy.  We're not there yet.